File: scripts/check_content_quality.py
#!/usr/bin/env python3
"""Validate Riftbook Markdown contributions with repository-specific rules."""

from __future__ import annotations

import argparse
import re
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable
from urllib.parse import parse_qsl, urlsplit

TRACKING_PREFIXES = ("utm_",)
TRACKING_KEYS = {"ref", "referrer", "affiliate", "aff", "campaign"}
PLACEHOLDER_RE = re.compile(r"\b(?:TBD|TODO|COMING SOON)\b", re.IGNORECASE)
MARKDOWN_LINK_RE = re.compile(r"!?\[[^\]]*]\(([^)]+)\)")
CODE_FENCE_RE = re.compile(r"```.*?```", re.DOTALL)
INLINE_CODE_RE = re.compile(r"`[^`\n]+`")
EXEMPT_FILES = {
    "CONTRIBUTING.md",
    "CURATION.md",
    "REVIEW.md",
    "ROADMAP.md",
    "CHANGELOG.md",
}
EXEMPT_PREFIXES = (
    "templates/",
    ".github/",
)


@dataclass(frozen=True)
class Finding:
    path: Path
    line: int
    rule: str
    message: str

    def render(self) -> str:
        return f"{self.path}:{self.line}: [{self.rule}] {self.message}"


def strip_code(text: str) -> str:
    text = CODE_FENCE_RE.sub("", text)
    return INLINE_CODE_RE.sub("", text)


def line_number(text: str, offset: int) -> int:
    return text.count("\n", 0, offset) + 1


def is_exempt(path: Path) -> bool:
    value = path.as_posix()
    return value in EXEMPT_FILES or value.startswith(EXEMPT_PREFIXES)


def tracking_parameters(url: str) -> list[str]:
    query = urlsplit(url).query
    result: list[str] = []
    for key, _ in parse_qsl(query, keep_blank_values=True):
        normalized = key.lower()
        if normalized.startswith(TRACKING_PREFIXES) or normalized in TRACKING_KEYS:
            result.append(key)
    return result


def validate_markdown(path: Path, text: str) -> list[Finding]:
    findings: list[Finding] = []
    scanned = strip_code(text)

    if not is_exempt(path):
        for match in PLACEHOLDER_RE.finditer(scanned):
            findings.append(
                Finding(
                    path,
                    line_number(scanned, match.start()),
                    "placeholder",
                    f"Remove placeholder text: {match.group(0)!r}.",
                )
            )

    for match in MARKDOWN_LINK_RE.finditer(scanned):
        target = match.group(1).strip().split(maxsplit=1)[0].strip("<>")
        if target.startswith(("http://", "https://")):
            params = tracking_parameters(target)
            if params:
                findings.append(
                    Finding(
                        path,
                        line_number(scanned, match.start(1)),
                        "tracking-parameter",
                        f"Remove tracking parameters from URL: {', '.join(params)}.",
                    )
                )

    if path.name != "README.md" and not is_exempt(path):
        if scanned.strip() and not scanned.lstrip().startswith("#"):
            findings.append(
                Finding(
                    path,
                    1,
                    "missing-title",
                    "Markdown content should start with an H1 title.",
                )
            )

    return findings


def discover_markdown(root: Path) -> list[Path]:
    return sorted(
        path
        for path in root.rglob("*.md")
        if ".git" not in path.parts and not any(part.startswith(".venv") for part in path.parts)
    )


def changed_markdown(root: Path, base: str) -> list[Path]:
    command = ["git", "diff", "--name-only", "--diff-filter=ACMR", f"{base}...HEAD", "--", "*.md"]
    completed = subprocess.run(command, cwd=root, check=False, text=True, capture_output=True)
    if completed.returncode != 0:
        raise RuntimeError(completed.stderr.strip() or "Unable to determine changed Markdown files.")
    return sorted(root / line for line in completed.stdout.splitlines() if line.strip())


def validate_paths(root: Path, paths: Iterable[Path]) -> list[Finding]:
    findings: list[Finding] = []
    for path in paths:
        absolute = path if path.is_absolute() else root / path
        if not absolute.exists():
            continue
        relative = absolute.relative_to(root)
        findings.extend(validate_markdown(relative, absolute.read_text(encoding="utf-8")))
    return findings


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=Path.cwd())
    parser.add_argument(
        "--changed-from",
        metavar="GIT_REF",
        help="Validate only Markdown files changed from the supplied Git ref.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    root = args.root.resolve()
    try:
        paths = changed_markdown(root, args.changed_from) if args.changed_from else discover_markdown(root)
    except RuntimeError as error:
        print(f"content-quality: {error}", file=sys.stderr)
        return 2

    findings = validate_paths(root, paths)
    for finding in findings:
        print(finding.render())

    if findings:
        print(f"\nFound {len(findings)} content quality issue(s) in {len(paths)} Markdown file(s).")
        return 1

    print(f"Content quality check passed for {len(paths)} Markdown file(s).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check-content-quality.mjs
#!/usr/bin/env node

import { readFile } from 'node:fs/promises';
import { resolve, relative, sep } from 'node:path';
import process from 'node:process';

const CONTENT_ROOTS = new Set([
  'learning',
  'playbook',
  'tools',
  'skills',
  'workflows',
  'frameworks',
  'prompts',
  'examples',
]);

const PLACEHOLDER_RE = /\b(?:TBD|TODO|coming soon)\b/i;
const MARKDOWN_LINK_RE = /\[[^\]]*\]\(([^)]+)\)/g;
const TRACKING_KEYS = new Set([
  'ref',
  'referrer',
  'affiliate',
  'aff',
  'campaign',
  'source',
]);

function normalizePath(file) {
  return relative(process.cwd(), resolve(file)).split(sep).join('/');
}

function isPublishedContent(file) {
  const normalized = normalizePath(file);
  const [root] = normalized.split('/');
  return normalized.endsWith('.md') && CONTENT_ROOTS.has(root);
}

function stripFencedCode(markdown) {
  const lines = markdown.split(/\r?\n/);
  let inFence = false;
  let fenceMarker = '';

  return lines.map((line) => {
    const match = line.match(/^\s*(```+|~~~+)/);
    if (match) {
      const marker = match[1][0];
      if (!inFence) {
        inFence = true;
        fenceMarker = marker;
      } else if (marker === fenceMarker) {
        inFence = false;
        fenceMarker = '';
      }
      return '';
    }
    return inFence ? '' : line;
  }).join('\n');
}

function firstMeaningfulLine(markdown) {
  const lines = markdown.replace(/^\uFEFF/, '').split(/\r?\n/);
  let inFrontmatter = false;
  let frontmatterSeen = false;

  for (const line of lines) {
    const trimmed = line.trim();
    if (!frontmatterSeen && trimmed === '---') {
      frontmatterSeen = true;
      inFrontmatter = true;
      continue;
    }
    if (inFrontmatter) {
      if (trimmed === '---') inFrontmatter = false;
      continue;
    }
    if (trimmed) return trimmed;
  }
  return '';
}

function hasTrackingParameter(destination) {
  const clean = destination.trim().replace(/^<|>$/g, '');
  if (!/^https?:\/\//i.test(clean)) return false;

  try {
    const url = new URL(clean);
    for (const key of url.searchParams.keys()) {
      const lowered = key.toLowerCase();
      if (lowered.startsWith('utm_') || TRACKING_KEYS.has(lowered)) return true;
    }
  } catch {
    return false;
  }
  return false;
}

export function inspectMarkdown(file, markdown) {
  if (!isPublishedContent(file)) return [];

  const findings = [];
  const visible = stripFencedCode(markdown);
  const firstLine = firstMeaningfulLine(markdown);

  if (!/^#\s+\S/.test(firstLine)) {
    findings.push({ rule: 'missing-h1', message: 'Published content must start with an H1 title.' });
  }

  visible.split(/\r?\n/).forEach((line, index) => {
    if (PLACEHOLDER_RE.test(line)) {
      findings.push({
        rule: 'placeholder',
        line: index + 1,
        message: 'Remove placeholder content such as TBD, TODO, or coming soon.',
      });
    }

    for (const match of line.matchAll(MARKDOWN_LINK_RE)) {
      if (hasTrackingParameter(match[1])) {
        findings.push({
          rule: 'tracking-url',
          line: index + 1,
          message: `Remove tracking or referral parameters from ${match[1]}`,
        });
      }
    }
  });

  return findings;
}

async function main() {
  const files = process.argv.slice(2).filter((arg) => arg.endsWith('.md'));
  if (files.length === 0) {
    console.log('No Markdown files supplied; nothing to validate.');
    return;
  }

  let failures = 0;
  for (const file of files) {
    if (!isPublishedContent(file)) continue;
    let markdown;
    try {
      markdown = await readFile(file, 'utf8');
    } catch (error) {
      console.error(`${file}: unable to read file: ${error.message}`);
      failures += 1;
      continue;
    }

    const findings = inspectMarkdown(file, markdown);
    for (const finding of findings) {
      const location = finding.line ? `:${finding.line}` : '';
      console.error(`${file}${location} [${finding.rule}] ${finding.message}`);
    }
    failures += findings.length;
  }

  if (failures > 0) {
    console.error(`Content quality failed with ${failures} finding(s).`);
    process.exitCode = 1;
  } else {
    console.log('Content quality passed.');
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  await main();
}
